### PR TITLE
[Indexing] Add `ConcatenateOp`

### PR DESCRIPTION
**Pull request by makslevental**
This is a wholesale lift/grab from [stablehlo](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#concatenate). Feel free to tell me you want more tests, or more examples in the td description, or to make a suggestion about a different asm.

Stack (bottom to top):

https://github.com/iree-org/iree-llvm-sandbox/pull/702

[here](https://github.com/iree-org/iree-llvm-sandbox/pull/696)

### CODE REVIEW

File: test/python/dialects/indexing/dialect.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def test_tensor_value():
   print(module)
 
   pm = PassManager.parse(
-      'builtin.module(func.func(convert-elementwise-to-linalg))')
+      "builtin.module(func.func(convert-elementwise-to-linalg))")
```
Comment (ingomueller-net) [inline test/python/dialects/indexing/dialect.py:94]: Why this change?

File: include/structured/Dialect/Indexing/IR/IndexingOps.td [hunk 2]
```diff
@@ -37,4 +37,33 @@ def Indexing_GatherOp : Indexing_Op<"gather", [
...
+
+  let arguments = (ins
+    Variadic<AnyRankedTensor>:$inputs,
+    I64Attr:$dimension
+  );
+
+  let results = (outs AnyRankedTensor);
+
+  let assemblyFormat = [{
+     custom<VariadicOperandWithAttribute>($inputs) `dim` `=` $dimension attr-dict
```
Comment (ingomueller-net) [inline include/structured/Dialect/Indexing/IR/IndexingOps.td:64]: Maybe you can change the assembly format slightly in order to avoid the custom directive. Maybe like this:

```suggestion
     `(` $inputs `)` `dim` `=` $dimension attr-dict
```

(Maybe add braces around `dim = ...`.)

File: include/structured/Dialect/Indexing/IR/Indexing.h [hunk 3]
```diff
@@ -18,7 +18,16 @@
 #include "structured/Dialect/Indexing/IR/IndexingOpsDialect.h.inc"
 
 namespace mlir {
-namespace indexing {} // namespace indexing
```
Comment (ingomueller-net) [inline include/structured/Dialect/Indexing/IR/Indexing.h:21]: I *think* that you shouldn't need these forward declarations (even if you needed the custom directive, see below). If it doesn't compile without these, you might miss an include to a generated header file...

**Review by ingomueller-net — APPROVED** [review]
LGTM!

File: test/python/dialects/indexing/dialect.py (hunk 1 above)
Comment (makslevental) [reply test/python/dialects/indexing/dialect.py:94]: either yapf or black did this. i'll undo it

File: include/structured/Dialect/Indexing/IR/IndexingOps.td (hunk 2 above)
Comment (makslevental) [reply include/structured/Dialect/Indexing/IR/IndexingOps.td:64]: i was trying to figure out how to get it to be something like this but didn't think of delimiting with parens (not much experience yet with the parser. thanks!
